[TypeChecker] SE-0213: Implement literal init via coercion

xedin opened the pull request:
Implementation is as follows: In `preCheckExpression` try to
detect if there is `T(literal)` call in the AST, replace it with
implicit `literal as T`, while trying to form type-checked AST,
after constraint solving, restore source information and drop unnecessary
coercion expression.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves: rdar://problem/17088188
Resolves: rdar://problem/39120081
Resolves: rdar://problem/23672697
Resolves: rdar://problem/40379985

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[AST] Make it so coercion expr use for literal init is distinguishable

xedin committed:
[TypeChecker] SE-0213: Implement literal init via coercion

Implementation is as follows: In `preCheckExpression` try to
detect if there is `T(literal)` call in the AST, replace it with
implicit `literal as T`, while trying to form type-checked AST,
after constraint solving, restore source information and drop
unnecessary coercion expression.

Resolves: rdar://problem/17088188
Resolves: rdar://problem/39120081
Resolves: rdar://problem/23672697
Resolves: rdar://problem/40379985

xedin committed:
[TypeChecker/SE-0213] NFC: Add couple of additional test-cases

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please benchmark

ravikandhadai commented on test/SILOptimizer/diagnostic_constant_propagation_floats_x86.swift:
```diff
@@ -109,8 +109,5 @@ func testIntToFloatConversion() {
-  // Note that the error here is because of an implicit conversion of the input
-  // literal to 'Int'.
-  _blackHole(Float80(18_446_744_073_709_551_617)) // expected-error {{integer literal '18446744073709551617' overflows when stored into 'Int'}}
+  _blackHole(Float80(18_446_744_073_709_551_617)) // Ok
```
@xedin Nice to see that the above FIXME's are addressed by this fix. Thanks!

jrose-apple commented on test/Sema/diag_erroneous_iuo.swift:
```diff
@@ -116,9 +116,9 @@ _ = Array<Int!>() // expected-error {{'!' is not allowed here; perhaps '?' was i
+_ = Optional<Int!>(nil) // expected-error 2 {{'!' is not allowed here; perhaps '?' was intended?}}{{17-18=?}}
 let _: Optional<Int!> = nil // expected-error {{'!' is not allowed here; perhaps '?' was intended?}}{{20-21=?}}
-_ = Int!?(0) // expected-error {{'!' is not allowed here; perhaps '?' was intended?}}{{8-9=?}}
+_ = Int!?(0) // expected-error 3 {{'!' is not allowed here; perhaps '?' was intended?}}{{8-9=?}}
```
This looks like we're processing the type multiple times…

xedin replied:
Yeah, we have that in a couple of places already unfortunately, because we don't track if we produced the same diagnostic multiple times...

slavapestov replied:
Looking at the TypeLoc, we should see it has already been resolved to a Type, and not resolve the TypeRepr again. Where is the diagnostic being emitted?

xedin replied:
Ah I see, it's probably my calls to resolveType which caused this, I'll try to fix it.

xedin replied:
Done, just needed to cache the type in a couple of places :)

xedin commented:
@swift-ci please benchmark

xedin commented:
@swift-ci please benchmark

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please smoke test compiler performance

rudkx reviewed (approved):
LGTM, thanks!

moiseev reviewed (approved):
stdlib change looks innocent. 👍

xedin commented:
@swift-ci please test

xedin commented:
Oops, looks like I missed one 32-bit test.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

DougGregor commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -1797,6 +1804,64 @@ void PreCheckExpression::resolveKeyPathExpr(KeyPathExpr *KPE) {
+    return nullptr;
+
+  // Don't bother to convert deprecated selector syntax.
+  if (auto selectorTy = TC.getObjCSelectorType(DC)) {
```
This is the use of string literals for selectors? I wonder if we can (separately) kill that now...

xedin replied:
That's right, it's `#selector(...)` vs. `Selector("...")`

DougGregor commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -3590,10 +3590,12 @@ ${assignmentOperatorComment(x.operator, True)}
-    let base = U${Self}(_value)
-    return self < (0 as ${Self}) ? ~base + 1 : base
+    @inline(__always)
+    get {
```
This seems unrelated?

xedin replied:
This is needed because constant propagation now looks at both branches, so as a way to work around that behavior @ravikandhadai and @moiseev told me that it would be appropriate to switch from `@_transparent` to `@inlineable`.

moiseev replied:
It is, unfortunately. `@_transparent` here causes the compile-time overflow on a branch that would not be taken at runtime.

ravikandhadai replied:
Just to clarify. The changes to how initializers are resolved, as a side-effect, makes constant propagation more precise as it can now propagate through more initializers. While in general it is a good thing, in this specific case, it resulted in a false positive due to mandatory inlining.

moiseev replied:
It is very much possible (and probably reasonable) to try the `&+` approach in a separate PR once this lands.

ravikandhadai replied:
@moiseev @xwu there is one other issue with "magnitude" (besides the +) that could trigger static errors if it is made @_transparent. It is passing a potentially negative (builtin Int) into a UInt : `let base = U${Self}(_value)`. This could trigger a static error on negative values like (-1).magnitude etc.  So perhaps keeping it @inline is better.

xwu replied:
No, `_value` is [the same type](https://github.com/xedin/swift/blob/29bbc99780265b6800f917f50f24f6818e9ae8e9/stdlib/public/core/Integers.swift.gyb#L3710) regardless of whether it's `Int` or `UInt`, and that initializer simply assigns `self._value = other._value`.

xedin commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -1797,6 +1804,64 @@ void PreCheckExpression::resolveKeyPathExpr(KeyPathExpr *KPE) {
+    auto *parent = ExprStack.back();
+    if (isa<BindOptionalExpr>(parent) || isa<ForceValueExpr>(parent))
+      return nullptr;
+  }
```
I was just trying to avoid breaking existing code like https://github.com/apple/swift/pull/17860/files#diff-de58b3c2866a24533bc39d1bfd40fc3cR276 but maybe I shouldn't do it at all then?

xedin replied:
Alright, this check is hidden under `!isSwiftVersionAtLeast(5)` now.

xedin committed:
[TypeChecker/SE-0213] Don't transform failable init calls into coercions in Swift < 5

Maintain source compatibility, but only when Swift < 5, in cases when
type declares conformance to literal protocol and at the same time has
failable initializer with the same parameter type, see `Source compatibility`
of the proposal for more details.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

DougGregor reviewed (approved):
LGTM, thanks!

xedin committed:
[TypeChecker/SE-0213] NFC: Fix one more improved constant propagation test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility
